### PR TITLE
Issue 3357

### PR DESCRIPTION
**Pull request by markiantorno**
Users were able to access data on partitions they did not have access to by using pagination links generated by users who did have access.

Fixes [issue 3357](https://github.com/hapifhir/hapi-fhir/issues/3357).

### CODE REVIEW

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3710-partition-based-access-control-is-not-enforced-on-opaque-pagination-links.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+type: fix
+issue: 2791
+jira: 3710
+title: "Users were able to access data on partitions they did not have access to by using pagination links generated by users who did have access."
```
Comment (tadgh) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3710-partition-based-access-control-is-not-enforced-on-opaque-pagination-links.yml:5]: This file needs to be renamed to `2791-partition-based-access.....` as the hapi-fhir issue # is the relevant one for post-processors.

Comment (tadgh) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_7_0/3710-partition-based-access-control-is-not-enforced-on-opaque-pagination-links.yml:5]:
```suggestion
title: "Users were previously able to access data on partitions they did not have access to by using pagination links generated by users who did have access."
```

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/BaseMultitenantResourceProviderR4Test.java [hunk 2]
```diff
@@ -109,6 +109,9 @@ private void createTenants() {
 	}
 
 	public void enableAuthorizationInterceptor(Supplier<List<IAuthRule>> theRuleSupplier) {
+		if(myAuthorizationInterceptor != null) {
+			ourRestServer.unregisterInterceptor(myAuthorizationInterceptor);
+		}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/BaseMultitenantResourceProviderR4Test.java:114]: Ah I see now why you made this change, but the new behaviour isn't necessarily "enable". I would rename this method so it is more accurate about what it is doing, maybe something like `setupAuthorizationInterceptorWithRules()`

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/AuthorizationInterceptorMultitenantJpaR4Test.java [hunk 3]
```diff
@@ -220,8 +225,51 @@ public void testSearchPagingAcrossTenants_Blocked() {
 			fail();
 		} catch (ForbiddenOperationException e) {
 			// good
+			System.out.println();
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/AuthorizationInterceptorMultitenantJpaR4Test.java:228]: remove

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/DatabaseBackedPagingProvider.java [hunk 4]
```diff
@@ -57,6 +60,7 @@ public DatabaseBackedPagingProvider(int theSize) {
 
 	@Override
 	public synchronized IBundleProvider retrieveResultList(RequestDetails theRequestDetails, String theId) {
+		myRequestPartitionHelperSvc.determineReadPartitionForRequestForSearchType(theRequestDetails, "Bundle", null, null);
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/search/DatabaseBackedPagingProvider.java:63]: Ah, so this does all the magic internally then? works for me.

**Review by tadgh — APPROVED** [review]
Approved pending minor changes. Also, pretty sure this is supposed to go into the release, so please re-point your PR to `rel_5_7` instead of master

**Comment by markiantorno** [comment]
this was already merged

**Pull request closed without merging**